spec/ddtrace/pin_spec.rb

```diff
@@ -94,32 +54,104 @@ def datadog_pin=(_pin)
+      it { expect(set_on[:a_setting]).to be(:a_value) }
+    end
+
+    context 'given an object with a Pin' do
+      context 'and a setting that does not conflict' do
+        let(:original_options) { { original_setting: :original_value } }
+
+        before { described_class.set_on(target, **original_options) }
+
+        it { is_expected.to be_a_kind_of(described_class) }
+        it { expect(set_on[:original_setting]).to be(:original_value) }
+        it { expect(set_on[:a_setting]).to be(:a_value) }
+      end
+
+      context 'and a setting that conflicts' do
+        let(:original_options) { { a_setting: :original_value } }
+
+        before { described_class.set_on(target, **original_options) }
+
+        it { is_expected.to be_a_kind_of(described_class) }
+        it { expect(set_on[:a_setting]).to be(:a_value) }
+      end
+
+      context 'which has a custom pin has already defined' do
+        let(:target_class) do
+          Class.new do
+            def datadog_pin
+              @custom_attribute
+            end
+
+            def datadog_pin=(_pin)
+              @custom_attribute = { custom_pin: true }
+            end
+          end
+        end
+
+        let(:target) { target_class.new }
+
+        before { pin.onto(target) }
+
+        it 'returns the custom pin' do
+          is_expected.to eq({ custom_pin: true }.merge(options))
+        end
+      end
+    end
+  end
+
+  describe '#initialize' do
+    before { pin }
+
+    context 'when given some options' do
+      let(:options) do
+        {
+          app: double('app'),
+          app_type: double('app_type'),
+          config: double('config'),
+          tags: double('tags'),
+        }
+      end
+
+      it do
+        expect(pin[:app]).to eq(options[:app])
+        expect(pin[:app_type]).to eq(options[:app_type])
+        expect(pin[:tags]).to eq(options[:tags])
+      end
+    end
+  end
+
+  describe '#onto' do
+    subject(:onto) { pin.onto(target) }
+    let(:returned_pin) { described_class.get_from(target) }
+
+    before { onto }
+
+    it 'attaches the pin to the target' do
+      expect(returned_pin).to be(pin)
+    end
+  end
+
   describe '#to_s' do
     subject(:string) { pin.to_s }
 
-    let(:service_name) { 'abc' }
     let(:options) { { app: 'anapp', app_type: 'db' } }
 
-    it { is_expected.to eq('Pin(service:abc,app:anapp,app_type:db,name:)') }
+    it { is_expected.to eq('Pin(app:anapp, app_type:db)') }
   end
 
   describe '#datadog_pin' do
     let(:returned_pin) { target.datadog_pin }
 
     before { pin.onto(target) }
 
-    it { expect(returned_pin.service).to eq(service_name) }
-  end
-
-  describe '#enabled?' do
-    subject(:enabled) { pin.enabled? }
-
-    it { is_expected.to be true }
-
-    context 'when the tracer is disabled' do
-      before { allow(Datadog::Tracing).to receive(:enabled?).and_return(false) }
-
-      it { is_expected.to be false }
-    end
+    it { expect(returned_pin).to be(pin) }
   end
 end
```

Comment:
Since we have non-trivial `#method_missing` and `#respond_to_missing?`, I suggest adding specs for these as well.

Reply:
I'm going to try to remove these.